packages/flutter_tools/lib/src/windows/build_windows.dart

```diff
@@ -102,16 +109,32 @@ Future<void> buildWindows(WindowsProject windowsProject, BuildInfo buildInfo, {
 }
 
 /// Writes the generatedPropertySheetFile with the configuration for the given build.
-void _writeGeneratedFlutterProperties(WindowsProject windowsProject, BuildInfo buildInfo, String target) {
+void _writeGeneratedFlutterProperties(
+  WindowsProject windowsProject,
+  BuildInfo buildInfo,
+  String target,
+) {
   final Map<String, String> environment = <String, String>{
     'FLUTTER_ROOT': Cache.flutterRoot,
     'FLUTTER_EPHEMERAL_DIR': windowsProject.ephemeralDirectory.path,
     'PROJECT_DIR': windowsProject.project.directory.path,
-    'TRACK_WIDGET_CREATION': (buildInfo?.trackWidgetCreation == true).toString(),
+    if (buildInfo.trackWidgetCreation != null)
+      'TRACK_WIDGET_CREATION': buildInfo.trackWidgetCreation.toString(),
+    if (buildInfo.treeShakeIcons != null)
+      'TREE_SHAKE_ICONS': buildInfo.treeShakeIcons.toString(),
+    if (buildInfo.extraGenSnapshotOptions?.isNotEmpty ?? false)
+      'EXTRA_GEN_SNAPSHOT_OPTIONS': buildInfo.extraGenSnapshotOptions.join(','),
+    if (buildInfo.extraFrontEndOptions?.isNotEmpty ?? false)
+      'EXTRA_FRONT_END_OPTIONS': buildInfo.extraFrontEndOptions.join(','),
+    if (buildInfo.dartDefines?.isNotEmpty ?? false)
+      'DART_DEFINES': buildInfo.dartDefines.join(','),
+    if (buildInfo.dartObfuscation != null)
+      'DART_OBFUSCATION': buildInfo.dartObfuscation.toString(),
+    if (buildInfo.splitDebugInfoPath != null)
+      'SPLIT_DEBUG_INFO': buildInfo.splitDebugInfoPath,
+    if (target != null)
+      'FLUTTER_TARGET': target,
```

Comment:
We should consider a helper that does all the buildInfo->environment dictionary parts of this, since I think at least the three desktop platforms will end up having identical code.

Reply:
Its slightly different for all three, since linux uses a makefile export and macOS uses the xcodeproj. There is definitely some refactoring we could do here to make it more sane, but I'll leave that for another day

Reply:
I know the way we publish the environment variables vary, it's the process of turning buildInfo into key/value pair that I was thinking we should collect in one place. Then each build system can do the trivial work of converting a dictionary of key/value pairs to the format of their choice (without any knowledge of what the keys and values actually are).